### ISSUE TITLE
why there isn't a more popular training approach based on exl2.

### ISSUE THREAD

**Issue by laoda513** [turboderp/exllamav2#309]
I'm a bit confused about the basis of your training. Currently, the fastest inference solution is exllamav2, but besides using the original gptq model, it's possible to directly train using autogptq and alpaca_lora_4bit, although it seems there's no way to directly train with exl2.
From what I see, mainstream training either involves training with the original model and then converting it to gptq/exl2, or using bnb 4bit to load the original model. Then, the lora weights trained and the converted gptq/exl2 model are loaded together for inference with exllamav2.
However, loading with bnb4bit/gptq4bit should result in the more loss of original weights than exl2 5/6bits. If lora weights trained on these weights are loaded together with the more accurate exl2, there should be some discrepancies. So, I'm quite curious why there isn't a more popular training approach based on exl2.

**Comment by turboderp** [comment]
The 53 GB vs 70 GB is just the difference in the weights. The 2-bit difference would be next to, say, 48 bits per parameter for gradients, and then you have large batches on top probably, maybe huge attention matrices if you're doing long-range finetuning.. that sort of thing.

**Comment by laoda513** [comment]
"Indeed. That's why I'm puzzled by the lack of attention to training based on the EXL2 format. During training, the required memory can increase significantly. For instance, using a 4-bit quantized training method like GPTQ, and applying gradient checkpoint or LoRA, it's possible to train the basic LLaMa2 70B model with 8x 3090 GPUs (I haven't tried long contexts yet). However, if training is based on BNB 8-bit, the memory requirements explode.

That's why I've always opted for training with GPTQ models.

When I saw that EXL2 supports LoRA, but there isn't a specific training method based on EXL2 format, I was curious: what methods do people use to train LoRA? Is it Transformer+BNB, GPTQ, direct training with original weights, or is there some other, better method?

I'm not an algorithm engineer, but based on my understanding of LoRA, its basic algorithm is based on W+deltaW. Even if deltaW is trained very well, different quantizations will inevitably cause some variations in W, and thus in W+deltaW. Could it be that the level of these differences doesn't lead to significant degradation in the combined model's performance, and that's why there hasn't been a specific focus on developing training methods based on the EXL2 format?"

**Comment by grimulkan** [comment]
I do think the 4-8 bit range can have a big impact on the training VRAM and therefore what is even accessible, so it would be worth it IMO. QLora doesn't really offer variable precision. It's not the overall size but rather hitting the ideal breakpoints, and a few GB per card can help there.

@turboderp I'm also curious how hard it would be to implement native auto-differentiation and backpropagation in Exllama, at least for LORA layers, and for decoder-only models like Llama. The math doesn't sound more complicated than what you're already doing I think (heck, the quantization code with Cholesky and inverses sounds way more numerically nasty than auto-diff). If the focus is on Llama-style architectures, you wouldn't need to support arbitrary compute graphs like pytorch auto-diff does, and it can be specialized to LORA layers. Guess you'd need a simple optimizer too (or plain SGD).

It'd be nice to have a training loop independent of all the crap and bloat that transformers + PEFT + Qlora/bnb etc. introduces, not to mention wrapper packages like deepspeed. For instance, there is no pipeline parallel training in conjunction with quantization anywhere, but it could be implemented more easily if the building blocks were present in a more low-level implementation like Exllama, without mucking with a nightmare of dependencies. I think a lot of bloat exists because a lot of the packages are highly generalized, and Llama (and Yi, Mistral, etc.) training options therefore suffer.
